br/pkg/restore/data.go

```diff
@@ -292,51 +300,52 @@ func (recovery *Recovery) WaitApply(ctx context.Context) (err error) {
-		if err := ectx.Err(); err != nil {
-			break
-		}
-		storeAddr := getStoreAddress(recovery.allStores, store.Id)
-		storeId := store.Id
-		workers.ApplyOnErrorGroup(eg, func() error {
-			recoveryClient, conn, err := recovery.newRecoveryClient(ectx, storeAddr)
-			if err != nil {
-				return errors.Trace(err)
-			}
-			defer conn.Close()
-			log.Info("resolve data to tikv", zap.String("tikv address", storeAddr), zap.Uint64("store id", storeId))
-			req := &recovpb.ResolveKvDataRequest{ResolvedTs: resolvedTs}
-			stream, err := recoveryClient.ResolveKvData(ectx, req)
-			if err != nil {
-				log.Error("send the resolve kv data failed", zap.Uint64("store id", storeId))
-				return errors.Trace(err)
-			}
-			// for a TiKV, received the stream
-			for {
-				var resp *recovpb.ResolveKvDataResponse
-				if resp, err = stream.Recv(); err == nil {
-					log.Info("current delete key", zap.Uint64("resolved key num", resp.ResolvedKeyCount), zap.Uint64("store id", resp.StoreId))
-				} else if err == io.EOF {
-					break
-				} else {
-					return errors.Trace(err)
-				}
-			}
-			recovery.progress.Inc()
-			log.Info("resolve kv data done", zap.String("tikv address", storeAddr), zap.Uint64("store id", storeId))
-			return nil
-		})
+	runner := rangetask.NewRangeTaskRunner("br-flashback-prepare-runner", recovery.mgr.GetStorage().(tikv.Storage), int(recovery.concurrency), handler)
+	// Run prepare flashback on the entire TiKV cluster. Empty keys means the range is unbounded.
+	err = runner.RunOnRange(ctx, []byte(""), []byte(""))
+	if err != nil {
+		log.Error("region flashback prepare get error")
+		return errors.Trace(err)
 	}
-	// Wait for all TiKV instances force leader and wait apply to last log.
-	return eg.Wait()
+
+	// TODO: currently, we do not know how many regions before start progress, EBS snapshot restore very fast at this stage (< 5mins), so we just simple add 1.
+	recovery.progress.Inc()
+	log.Info("region flashback prepare complete", zap.Int("regions", runner.CompletedRegions()))
+
+	return nil
+}
+
+// flashback the region data to version resolvedTs
+func (recovery *Recovery) FlashbackToVersion(ctx context.Context, resolvedTs uint64, commitTs uint64) (err error) {
+	handler := func(ctx context.Context, r tikvstore.KeyRange) (rangetask.TaskStat, error) {
+		return ddl.SendFlashbackToVersionRPC(ctx, recovery.mgr.GetStorage().(tikv.Storage), resolvedTs, commitTs-1, commitTs, r)
+	}
+
+	runner := rangetask.NewRangeTaskRunner("br-flashback-runner", recovery.mgr.GetStorage().(tikv.Storage), int(recovery.concurrency), handler)
+	// Run flashback on the entire TiKV cluster. Empty keys means the range is unbounded.
+	err = runner.RunOnRange(ctx, []byte(""), []byte(""))
+	if err != nil {
+		log.Error("region flashback get error",
+			zap.Uint64("resolvedTs", resolvedTs),
+			zap.Uint64("commitTs", commitTs),
+			zap.Int("regions", runner.CompletedRegions()))
+		return errors.Trace(err)
+	}
+
+	recovery.progress.Inc()
```

Comment:
Could get region count by rangetask.TaskStat in the prepare phase and show progress in `FlashbackToVersion`.